Use the presence of hover styles as a signal for InteractionRegion consolidation

etiennesegonzac opened the pull request:
#### 969122115609f420b20ac90102f61e4e934a0af3
<pre>
Use the presence of hover styles as a signal for InteractionRegion consolidation
<a href="https://bugs.webkit.org/show_bug.cgi?id=259398">https://bugs.webkit.org/show_bug.cgi?id=259398</a>
&lt;rdar://111551476&gt;

Reviewed by Tim Horton.

Relax the consolidation rule when the ancestor has hover styles, so
that &quot;buttons&quot; with child spans get a single InteractionRegion.

* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::elementMatchesHoverRules):
Export `elementMatchesHoverRules`.

* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::shouldConsolidateInteractionRegion):
Allow consolidation if the ancestor has hover rules and the element is
centered in either axis.

* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions.html:
Add a test covering the change.

Canonical link: <a href="https://commits.webkit.org/266230@main">https://commits.webkit.org/266230@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52482ec3df1c27eb08a7958ac1066a9821f2f2cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15018 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12642 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13360 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13621 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15329 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13448 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15636 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11979 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19038 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12469 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12146 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15364 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12648 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10506 "1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11919 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11843 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3240 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16241 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12490 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -227,7 +227,7 @@ void updateLayersForInteractionRegions(const RemoteLayerTreeNode& node)
                     [regionLayer setBorderColor:cachedCGColor({ WebCore::SRGBA<float>(1, 0, 0, .2) }).get()];
                     [regionLayer setBorderWidth:6];
-                    [regionLayer setName:@"Occlusion"];
+                    [regionLayer setName:[NSString stringWithFormat:@"Occlusion %llu", region.elementIdentifier.toUInt64()]];
```
Is this worth the extra cost per layer (of which there can be many)? Also I wonder if it would be cheaper to use WTF string building stuff instead of format strings.

etiennesegonzac replied:
moving this change to a future patch where it'll be more on-topic

hortont424 commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -197,18 +197,24 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
-            && !renderer.hasVisibleBoxDecorations();
+            && marginBottom <= maxMargin;
+
+        constexpr auto centeredWiggleRoom = 2;
```
Not sure about "wiggle room" in the tree (though I love it personally).

etiennesegonzac replied:
it is in the tree! but only in tests :D 
went with `offCenterThreshold`

etiennesegonzac committed:
Use the presence of hover styles as a signal for InteractionRegion consolidation
https://bugs.webkit.org/show_bug.cgi?id=259398
<rdar://111551476>

Reviewed by Tim Horton.

Relax the consolidation rule when the ancestor has hover styles, so
that "buttons" with child spans get a single InteractionRegion.

* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::elementMatchesHoverRules):
Export `elementMatchesHoverRules`.

* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::shouldConsolidateInteractionRegion):
Allow consolidation if the ancestor has hover rules and the element is
centered in either axis.

* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions.html:
Add a test covering the change.

Canonical link: https://commits.webkit.org/266230@main
